assets/js/modules/reader-revenue-manager/components/common/PublicationOnboardingStateNotice.test.js

```diff
@@ -144,4 +145,49 @@ describe( 'PublicationOnboardingStateNotice', () => {
 			);
 		}
 	);
+
+	it( 'should sync onboarding state when the window is refocused 15 seconds after clicking the CTA', () => {
+		jest.useFakeTimers();
+
+		mockSyncPublication = jest.spyOn(
+			registry.dispatch( MODULES_READER_REVENUE_MANAGER ),
+			'syncPublicationOnboardingState'
+		);
+		mockSyncPublication.mockImplementation( () => Promise.resolve() );
+
+		registry
+			.dispatch( MODULES_READER_REVENUE_MANAGER )
+			.receiveGetSettings( {
+				publicationID: 'ABCDEFGH',
+				publicationOnboardingState: ONBOARDING_ACTION_REQUIRED,
+				publicationOnboardingStateLastSyncedAtMs: 0,
+			} );
+
+		const { getByText } = render( <PublicationOnboardingStateNotice />, {
+			registry,
+			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
+		} );
+
+		act( () => {
+			fireEvent.click( getByText( 'Complete publication setup' ) );
+		} );
+
+		act( () => {
+			global.window.dispatchEvent( new Event( 'blur' ) );
+		} );
+
+		expect( mockSyncPublication ).toHaveBeenCalledTimes( 0 );
+
+		act( () => {
+			jest.advanceTimersByTime( 15000 );
+		} );
+
+		act( () => {
+			global.window.dispatchEvent( new Event( 'focus' ) );
+		} );
+
+		expect( mockSyncPublication ).toHaveBeenCalledTimes( 1 );
```

Comment:
Although we have started using mocks a bit more lately, I think we should still aim to test in a more integrated manner where it's practical to do so.

This test, and the similar one for `RRMSetupSuccessSubtleNotification` look like they could be written using `fetchMock` to provide a response for the `GET publications` endpoint and then verifying the call has been made, the onboarding state has been updated, and the component has re-rendered accordingly.

I'd suggest taking this route, unless this provides excessively complicated to set up.

Reply:
I absolutely agree with integrated tests instead of just mocking functionality. I had actually tried using `fetchMock` here, but I couldn't make it work to the end as I wasn't able to exactly mock the `POST` request to RRM settings endpoint, most likely related to the timestamp passed for `publicationOnboardingStateLastSyncedAtMs`. I tried to bypass that by using `muteFetch`, but unfortunately, that wasn't very fruitful as well. Considering the complexities around waiting times and usage of fake timers here, do you think it is satisfactory here to go with the mock?

Reply:
Hmm, I had a look myself and could see where you ran into some difficulties - I think the tricky bit is that the POST to the settings endpoint ends up being made after the test has cleaned up the `fetchMock` handlers unless we wait for it. Using `waitFor()` solves this, though, and the test can be written in an integrated way as follows:

```diff
diff --git a/assets/js/modules/reader-revenue-manager/components/common/PublicationOnboardingStateNotice.test.js b/assets/js/modules/reader-revenue-manager/components/common/PublicationOnboardingStateNotice.test.js
index f5f7681c41..7bb73724f2 100644
--- a/assets/js/modules/reader-revenue-manager/components/common/PublicationOnboardingStateNotice.test.js
+++ b/assets/js/modules/reader-revenue-manager/components/common/PublicationOnboardingStateNotice.test.js
@@ -26,6 +26,7 @@ import {
 	provideUserAuthentication,
 	provideUserInfo,
 	render,
+	waitFor,
 } from '../../../../../../tests/js/test-utils';
 import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../googlesitekit/constants';
 import * as tracking from '../../../../util/tracking';
@@ -34,6 +35,7 @@ import {
 	PUBLICATION_ONBOARDING_STATES,
 } from '../../datastore/constants';
 import PublicationOnboardingStateNotice from './PublicationOnboardingStateNotice';
+import * as fixtures from '../../datastore/__fixtures__';
 
 const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
 mockTrackEvent.mockImplementation( () => Promise.resolve() );
@@ -146,28 +148,50 @@ describe( 'PublicationOnboardingStateNotice', () => {
 		}
 	);
 
-	it( 'should sync onboarding state when the window is refocused 15 seconds after clicking the CTA', () => {
-		jest.useFakeTimers();
+	it( 'should sync onboarding state when the window is refocused 15 seconds after clicking the CTA', async () => {
+		const originalDateNow = Date.now;
+
+		// Mock the date to be an arbitrary time.
+		const mockNow = new Date( '2020-01-01 12:30:00' ).getTime();
+		Date.now = jest.fn( () => mockNow );
+
+		const publicationsEndpoint = new RegExp(
+			'^/google-site-kit/v1/modules/reader-revenue-manager/data/publications'
+		);
 
-		mockSyncPublication = jest.spyOn(
-			registry.dispatch( MODULES_READER_REVENUE_MANAGER ),
-			'syncPublicationOnboardingState'
+		const settingsEndpoint = new RegExp(
+			'^/google-site-kit/v1/modules/reader-revenue-manager/data/settings'
 		);
-		mockSyncPublication.mockImplementation( () => Promise.resolve() );
+
+		jest.useFakeTimers();
 
 		registry
 			.dispatch( MODULES_READER_REVENUE_MANAGER )
 			.receiveGetSettings( {
-				publicationID: 'ABCDEFGH',
+				publicationID: 'QRSTUVWX',
 				publicationOnboardingState: ONBOARDING_ACTION_REQUIRED,
 				publicationOnboardingStateLastSyncedAtMs: 0,
 			} );
 
-		const { getByText } = render( <PublicationOnboardingStateNotice />, {
-			registry,
-			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
+		fetchMock.getOnce( publicationsEndpoint, {
+			body: fixtures.publications,
+			status: 200,
 		} );
 
+		fetchMock.postOnce( settingsEndpoint, ( url, opts ) => {
+			const { data } = JSON.parse( opts.body );
+			// Return the same settings passed to the API.
+			return { body: data, status: 200 };
+		} );
+
+		const { container, getByText } = render(
+			<PublicationOnboardingStateNotice />,
+			{
+				registry,
+				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
+			}
+		);
+
 		act( () => {
 			fireEvent.click( getByText( 'Complete publication setup' ) );
 		} );
@@ -176,8 +200,6 @@ describe( 'PublicationOnboardingStateNotice', () => {
 			global.window.dispatchEvent( new Event( 'blur' ) );
 		} );
 
-		expect( mockSyncPublication ).toHaveBeenCalledTimes( 0 );
-
 		act( () => {
 			jest.advanceTimersByTime( 15000 );
 		} );
@@ -186,8 +208,28 @@ describe( 'PublicationOnboardingStateNotice', () => {
 			global.window.dispatchEvent( new Event( 'focus' ) );
 		} );
 
-		expect( mockSyncPublication ).toHaveBeenCalledTimes( 1 );
+		await waitFor( () => {
+			expect( fetchMock ).toHaveFetched( publicationsEndpoint );
+			expect( fetchMock ).toHaveFetched( settingsEndpoint, {
+				body: {
+					data: {
+						publicationID: 'QRSTUVWX',
+						publicationOnboardingState: ONBOARDING_COMPLETE,
+						publicationOnboardingStateLastSyncedAtMs: Date.now(),
+					},
+				},
+			} );
+		} );
+
+		expect(
+			registry
+				.select( MODULES_READER_REVENUE_MANAGER )
+				.getPublicationOnboardingState()
+		).toBe( ONBOARDING_COMPLETE );
+
+		expect( container ).toBeEmptyDOMElement();
 
-		mockSyncPublication.mockReset();
+		// Restore Date.now method.
+		Date.now = originalDateNow;
 	} );
 } );
```

I think this approach is manageable, so we might as well stick with the integrated approach here. What do you reckon?

Reply:
I see what you did with the fetch mock for the settings endpoint, very clever! Thank you for pointing out this approach. The test cases have been updated accordingly.